vocode/streaming/agent/action_agent.py

```diff
@@ -0,0 +1,113 @@
+import logging
+from typing import AsyncGenerator, Optional, Tuple
+import re
+import openai
+
+
+from langchain.chat_models import ChatOpenAI
+from langchain.memory import ConversationBufferMemory
+from vocode import getenv
+from vocode.streaming.action.factory import ActionFactory
+from vocode.streaming.action.nylas_send_email import NylasSendEmail
+from vocode.streaming.agent.base_agent import BaseAgent
+from vocode.streaming.agent.prompts.action_prompt import ACTION_PROMPT_DEFAULT
+from vocode.streaming.models.actions import ActionType
+from vocode.streaming.models.agent import ActionAgentConfig
+
+from langchain.prompts import (
+    ChatPromptTemplate,
+    MessagesPlaceholder,
+    SystemMessagePromptTemplate,
+    HumanMessagePromptTemplate,
+)
+from langchain.chains import ConversationChain
+
+from vocode.streaming.models.message import BaseMessage
+
+
+class ActionAgent(BaseAgent):
+    def __init__(
+        self, agent_config: ActionAgentConfig, logger: Optional[logging.Logger] = None
+    ):
+        super().__init__(agent_config, logger)
+        self.agent_config = agent_config
+        self.action_chain = []
+
+        openai.api_key = getenv("OPENAI_API_KEY")
+        if not openai.api_key:
+            raise ValueError("OPENAI_API_KEY must be set in environment or passed in")
+        self.prompt = ChatPromptTemplate.from_messages(
+            [
+                SystemMessagePromptTemplate.from_template(
+                    ACTION_PROMPT_DEFAULT.format(actions=self.get_action_descriptions())
+                ),
+                MessagesPlaceholder(variable_name="history"),
+                HumanMessagePromptTemplate.from_template("{input}"),
+            ]
+        )
+        self.memory = ConversationBufferMemory(return_messages=True)
+
+        self.llm = ChatOpenAI(
+            model_name=self.agent_config.model_name,
+            temperature=self.agent_config.temperature,
+            max_tokens=self.agent_config.max_tokens,
+            openai_api_key=openai.api_key,
+        )
+        self.conversation = ConversationChain(
+            memory=self.memory, prompt=self.prompt, llm=self.llm
+        )
+        self.is_first_response = True
+
+    async def respond(
+        self,
+        human_input,
+        is_interrupt: bool = False,
+        conversation_id: Optional[str] = None,
+    ) -> Tuple[str, bool]:
+        if self.is_first_response and self.agent_config.initial_message:
+            response = self.agent_config.initial_message
+            self.is_first_response = False
+        else:
+            verbose_response = await self.conversation.apredict(input=human_input)
+            response = self.extract_response(verbose_response)
+            self.run_actions(verbose_response)
+
+        self.logger.debug(f"LLM verbose response: {verbose_response}")
+        return response, False
+
+    def get_action_descriptions(self):
+        descriptions = []
+        action_factory = ActionFactory()
```

Comment:
To be honest, I would prefer if this action factory is passed as an argument to this agent.
Would make it easier to switch between action factories